### ISSUE TITLE
Error when modifying element of the list 

### ISSUE THREAD

**Issue by silenc3r** [microsoft/pyright#2132]
This code
```
class Klass:
    def __init__(self):
        self.table = [None] * 10
        self.table[0] = 0
```
gives following error (on last line):
```
[Pyright reportGeneralTypeIssues] [E] No overloads for "__setitem__" match the provided arguments Argument types: (Literal[0], Literal[0])
```

**Comment by erictraut** [comment]
This is the correct behavior. You are assigning the expression `[None] * 10`, which is inferred to be of type `List[None]`, but then you are assigning an `int` to the list, which is incompatible with `List[None]`.

The correct solution here is to provide a type annotation that indicates that your list is actually of type `List[Optional[int]]`.

```python
class Klass:
    def __init__(self):
        self.table: List[Optional[int]] = [None] * 10
        self.table[0] = 0
```

For a better understanding of type inference and why type annotations are needed in some cases to eliminate ambiguity for the type checker, please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md).
